src/SpannerClient/SpannerClientFactory.php

```diff
@@ -33,51 +35,74 @@ class SpannerClientFactory
     private const SESSIONS_MIN = 1;
     private const SESSIONS_MAX = 100;
 
+    /** @var CacheItemPoolInterface|null */
+    private $authCache;
+
+    /** @var array */
+    private $configuration;
+
+    /** @var string|null */
+    private $keyFilePath;
+
+    public function __construct(
+        CacheItemPoolInterface $authCache = null,
+        array $configuration = null,
+        string $keyFilePath = null
+    ) {
+        $this->authCache = $authCache ?? new SysVCacheItemPool();
+        $this->configuration = $configuration ?? [];
+        $this->keyFilePath = $keyFilePath === null
+            ? ($_ENV[self::KEY_FILE_ENV_VARIABLE] ?? self::DEFAULT_CREDENTIALS_FILE)
+            : $keyFilePath;
+    }
+
     /**
-     * Creates a Google Spanner client from env configuration.
-     *
-     * @return SpannerClient
-     * @throws GoogleException When ext/grpc is missing.
+     * @throws GoogleException
      */
-    public function create()
+    public function create(): SpannerClient
     {
-        $authCache = new SysVCacheItemPool();
-        $keyFileName = $_ENV[self::KEY_FILE_ENV_VARIABLE] ?? self::DEFAULT_CREDENTIALS_FILE;
-        if ($keyFileName === '') {
-            new GoogleException(
-                sprintf(
-                    'Missing path to Google credentials key file (should be set as an environment variable "%s").',
-                    self::KEY_FILE_ENV_VARIABLE
-                )
-            );
-        }
-
-        $keyFile = json_decode(file_get_contents($keyFileName), true);
-
-        return new SpannerClient(['keyFile' => $keyFile, 'authCache' => $authCache]);
+        return new SpannerClient(
+            array_merge(
+                [
+                    'keyFile' => $this->getKeyFileParsedContent(),
+                    'authCache' => $this->authCache
+                ],
+                $this->configuration
+            )
+        );
     }
 
     /**
-     * Creates a session pool to allow multiple sessions to share the auth cache.
+     * @TODO We must remove this method in the next major version
      *
-     * @return CacheSessionPool
-     * @throws \Exception
+     * @deprecated Please DO NOT use this method. The client of this library should define it if needed
```

Comment:
What prevents from removing deprecated  code from current version?

Reply:
We already received feedback from management that we are increasing major version in an unplanned way. We should deprecate code, plan a major version and do it. So I would prefer go this way. It is also a good habit, specially for extensions (which is not the case now, I know).